pyro/distributions/sine_bivariate_von_mises.py

```diff
@@ -0,0 +1,280 @@
+    avoid parameterizations where the distribution becomes bimodal; see note below.
+
+    .. note:: Sample efficiency drops as
+
+        .. math::
+
+            \frac{\rho}{\kappa_1\kappa_2} \rightarrow 1
+
+        because the distribution becomes increasingly bimodal.
+
+    .. note:: The correlation and weighted_correlation params are mutually exclusive.
+
+    .. note:: This distribution does not work with :class:`~pyro.infer.svi.SVI`.
+
+    ** References: **
+      1. Probabilistic model for two dependent circular variables Singh, H., Hnizdo, V., and Demchuck, E. (2002)
+
+    :param torch.Tensor phi_loc: location of first angle
+    :param torch.Tensor psi_loc: location of second angle
+    :param torch.Tensor phi_concentration: concentration of first angle
+    :param torch.Tensor psi_concentration: concentration of second angle
+    :param torch.Tensor correlation: correlation between the two angles
+    :param torch.Tensor weighted_correlation: set correlation to weigthed_corr * sqrt(phi_conc*psi_conc)
+        to avoid bimodality (see note).
+    """
+
+    arg_constraints = {'phi_loc': constraints.real, 'psi_loc': constraints.real,
+                       'phi_concentration': constraints.positive, 'psi_concentration': constraints.positive,
+                       'correlation': constraints.real}
+    support = constraints.independent(constraints.real, 1)
+    max_sample_iter = 1000
+
+    def __init__(self, phi_loc, psi_loc, phi_concentration, psi_concentration, correlation=None,
+                 weighted_correlation=None, validate_args=None):
+
+        assert (correlation is None) != (weighted_correlation is None)
+
+        if weighted_correlation is not None:
+            correlation = weighted_correlation * (phi_concentration * psi_concentration).sqrt() + 1e-8
+
+        if torch.any(phi_concentration * psi_concentration <= correlation ** 2):
+            warnings.warn(
+                f'{self.__class__.__name__} bimodal due to concentration-correlation relation, '
+                f'sampling will likely fail.', UserWarning)
+
+        phi_loc, psi_loc, phi_concentration, psi_concentration, correlation = broadcast_all(phi_loc, psi_loc,
+                                                                                            phi_concentration,
+                                                                                            psi_concentration,
+                                                                                            correlation)
+        self.phi_loc = phi_loc
+        self.psi_loc = psi_loc
+        self.phi_concentration = phi_concentration
+        self.psi_concentration = psi_concentration
+        self.correlation = correlation
+        self._norm_const = None
+        event_shape = torch.Size([2])
+        batch_shape = phi_loc.shape
+        super(SineBivariateVonMises, self).__init__(batch_shape, event_shape, validate_args)
+
+    @property
```

Comment:
I would recommend using `@lazy_property` here, so as to use existing distributions idioms.  That will also allow you to remove the `self._norm_const` member:
```py
from torch.distributions.utils import lazy_property

class SineBivariateVonMises(TorchDistribution):
    ...
    @lazy_property
    def norm_const(self):
        corr = self.correlation.view(1, -1) + 1e-8
        ...
        return norm_const.reshape(self.phi_loc.shape)
```

Reply:
I see, thanks. I'll change this.